q2_types/tabular/_deferred_setup/_transformers.py

```diff
@@ -52,6 +301,22 @@ def table_jsonl_header(df: pd.DataFrame) -> str:
     return json.dumps(header, separators=(',', ':'))
 
 
+@plugin.register_transformer
+def df_to_table_jsonl(obj: pd.DataFrame) -> TableJSONLFileFormat:
+    obj = _copy_dataframe_with_attrs(obj)
+
+    header = table_jsonl_header(obj)
```

Comment:
I would like to suggest that `table_jsonl_header` does not mutate the dataframe, I think this can be accomplished by separating out the conversion logic into it's own helper `_standardize_dtypes()` which may or may not include `_copy_dataframe_with_attrs`. From that point, we could assume that `.attrs['type']` is set so that `table_jsonl_header` can read it.

I think this just looks like copying the for-loop from `table_jsonl_header` and moving that into its own method which more clearly mutates the copy of the dataframe.